include/dcsam/DCFactor.h

```diff
@@ -248,22 +252,63 @@ class DCFactor : public gtsam::Factor {
    */
   std::vector<double> evalProbs(const gtsam::DiscreteKey& dk,
                                 const gtsam::Values& continuousVals) const {
-    double total = 0.0;
-    std::vector<double> probs;
+    /*
+     * Normalizing a set of log probabilities in a numerically stable way is
+     * tricky. To avoid overflow/underflow issues, we compute the largest
+     * (finite) log probability and subtract it from each log probability before
+     * normalizing. This comes from the observation that if:
+     *    p_i = exp(L_i) / ( sum_j exp(L_j) ),
+     * Then,
+     *    p_i = exp(Z) exp(L_i - Z) / (exp(Z) sum_j exp(L_j - Z)),
+     *        = exp(L_i - Z) / ( sum_j exp(L_j - Z) )
+     *
+     * Setting Z = max_j L_j, we can avoid numerical issues that arise when all
+     * of the (unnormalized) log probabilities are either very large or very
+     * small.
+     */
+    std::vector<double> logProbs;
+    double maxLogProb = -std::numeric_limits<double>::infinity();
     for (size_t i = 0; i < dk.second; i++) {
       DiscreteValues testDiscreteVals;
       testDiscreteVals[dk.first] = i;
-      // Compute _unnormalized_ probability of discrete value = i
       // Recall: `error` returns -log(prob), so we compute exp(-error) to
       // recover probability
-      double probPrime = exp(-error(continuousVals, testDiscreteVals));
-      probs.push_back(probPrime);
+      double logProb = -error(continuousVals, testDiscreteVals);
+      logProbs.push_back(logProb);
+      if ((logProb != std::numeric_limits<double>::infinity()) &&
+          logProb > maxLogProb) {
+        maxLogProb = logProb;
+      }
+    }
+
+    // After computing the max = "Z" of the log probabilities L_i, we compute
+    // the log of the normalizing constant, log S, where S = sum_j exp(L_j - Z).
+    double total = 0.0;
+    for (size_t i = 0; i < dk.second; i++) {
+      double probPrime = exp(logProbs[i] - maxLogProb);
       total += probPrime;
     }
+    double logTotal = log(total);
 
-    // Normalize the result
+    // Now we compute the (normalized) probability (for each i):
+    // p_i = exp(L_i - Z - log S)
+    double checkNormalization = 0.0;
+    std::vector<double> probs;
     for (size_t i = 0; i < dk.second; i++) {
-      probs[i] = probs[i] / total;
+      double prob = exp(logProbs[i] - maxLogProb - logTotal);
+      probs.push_back(prob);
+      checkNormalization += prob;
+    }
+
+    if (checkNormalization != !.0) {
```

Comment:
Typo, I think this should be 1.0

Reply:
Holy cow - great catch! Not sure how this happened...